### PR TITLE
fix closed accepting Channels in closed ServerSocketChannels

### PR DESCRIPTION
**Pull request by weissi**
many thanks to @vlm for reporting 💯

Motivation:

For servers NIO had a race where when you start tearing down the
`ServerSocketChannel` and at roughly the same time a client connects, we
would not fully bring that channel up, nor tear it down. That lead to
open channels being deallocated which leads to a crash as that's an
illegal state.

Modifications:

Made sure newly accepted channels get closed if the
`ServerSocketChannel` is already clsing.

Result:

No crashes anymore when a client connects to a closing server.

### CODE REVIEW

File: Tests/NIOTests/ChannelTests.swift [hunk 1]
```diff
@@ -1526,4 +1527,106 @@ public class ChannelTests: XCTestCase {
...
+                }
+            }
+        }
+
+        do {
+            for _ in 0..<1000 {
+                try runTest()
+            }
+        } catch {
+            XCTFail("unexpected error: \(error)")
```
[the remark this answers is not in the document]
Comment (weissi) [reply Tests/NIOTests/ChannelTests.swift]: @normanmaurer I can if you want but the error message is way better this way. If I just let it throw you just get `ChannelError, code: 1` :|

[the remark this answers is not in the document]
Comment (normanmaurer) [reply Tests/NIOTests/ChannelTests.swift]: ok... then just leave it I guess

**Comment by normanmaurer** [comment]
@weissi please rebase

**Comment by weissi** [comment]
@Lukasa agreed. I'm happy to try writing a deterministic one as a separate PR. Do file an issue against me doing that...

**Comment by normanmaurer** [comment]
@weissi @vlm thanks a lot!

**Pull request merged**